File: src/lib/notion-project-fields.ts
import type { PageObjectResponse } from "@notionhq/client/build/src/api-endpoints.js";
import { config } from "../config.js";

/**
 * Notion project database property names (case-sensitive). Adjust if yours differ.
 */
export const PROJECT_PROPS = {
  title: "Name",
  status: "Status",
  deadline: "Deadline",
  priority: "Priority",
  currentFocus: "Current focus",
  nextAction: "Next action",
} as const;

export type ProjectBriefFields = {
  name: string;
  status: string;
  priority: string;
  currentFocus: string;
  nextAction: string;
  deadline: string | null;
};

function readRichText(prop: unknown): string {
  if (!prop || typeof prop !== "object" || !("type" in prop)) return "";
  const p = prop as {
    type: string;
    rich_text?: { plain_text: string }[];
    title?: { plain_text: string }[];
  };
  if (p.type === "rich_text" && Array.isArray(p.rich_text)) {
    return p.rich_text.map((t) => t.plain_text).join("");
  }
  if (p.type === "title" && Array.isArray(p.title)) {
    return p.title.map((t) => t.plain_text).join("");
  }
  return "";
}

function readPriority(prop: unknown): string {
  if (!prop || typeof prop !== "object" || !("type" in prop)) return "";
  const p = prop as {
    type: string;
    select?: { name: string } | null;
    status?: { name: string } | null;
    number?: number | null;
    rich_text?: { plain_text: string }[];
  };
  if (p.type === "select") return p.select?.name ?? "";
  if (p.type === "status") return p.status?.name ?? "";
  if (p.type === "number" && p.number != null && !Number.isNaN(p.number)) return String(p.number);
  if (p.type === "rich_text" && p.rich_text) {
    return p.rich_text.map((t) => t.plain_text).join("");
  }
  return "";
}

function titleFromPage(page: PageObjectResponse, titlePropName: string): string {
  const named = page.properties[titlePropName];
  if (named?.type === "title") {
    return named.title.map((t) => t.plain_text).join("") || "Untitled";
  }
  for (const [, prop] of Object.entries(page.properties)) {
    if (
      prop &&
      typeof prop === "object" &&
      "type" in prop &&
      prop.type === "title" &&
      "title" in prop
    ) {
      const title = (prop as { title: { plain_text: string }[] }).title;
      return title.map((t) => t.plain_text).join("") || "Untitled";
    }
  }
  return "Untitled";
}

/** Plain-text snapshot for briefs, tools, and LLM context (no Notion URLs). */
export function extractProjectBriefFields(page: PageObjectResponse): ProjectBriefFields {
  const props = page.properties;
  let status = "";
  const st = props[PROJECT_PROPS.status];
  if (st?.type === "status") status = st.status?.name ?? "";
  else if (st?.type === "select") status = st.select?.name ?? "";

  let deadline: string | null = null;
  const dl = props[PROJECT_PROPS.deadline];
  if (dl?.type === "date") deadline = dl.date?.start ?? null;

  return {
    name: titleFromPage(page, PROJECT_PROPS.title),
    status,
    priority: readPriority(props[PROJECT_PROPS.priority]),
    currentFocus: readRichText(props[PROJECT_PROPS.currentFocus]),
    nextAction: readRichText(props[PROJECT_PROPS.nextAction]),
    deadline,
  };
}

/** Database query filter for Status / single-select by kind from env. */
export function notionStatusDbFilter(
  propertyName: string,
  equals: string
): { property: string; select: { equals: string } } | { property: string; status: { equals: string } } {
  if (config.NOTION_STATUS_PROPERTY_KIND === "select") {
    return { property: propertyName, select: { equals } };
  }
  return { property: propertyName, status: { equals } };
}

/** `pages.create` / `pages.update` payload for status or select column. */
export function notionStatusUpdatePayload(name: string) {
  if (config.NOTION_STATUS_PROPERTY_KIND === "select") {
    return { select: { name } };
  }
  return { status: { name } };
}

export function notionStatusClearPayload() {
  if (config.NOTION_STATUS_PROPERTY_KIND === "select") {
    return { select: null };
  }
  return { status: null };
}

File: src/lib/notion_project_fields.py
from typing import Any

from src.config import get_settings

PROJECT_PROPS = {
    "title": "Name",
    "status": "Status",
    "deadline": "Deadline",
    "priority": "Priority",
    "currentFocus": "Current focus",
    "nextAction": "Next action",
}


def read_rich_text(prop: object) -> str:
    if not isinstance(prop, dict):
        return ""
    if prop.get("type") == "rich_text":
        return "".join(t.get("plain_text", "") for t in prop.get("rich_text") or [])
    if prop.get("type") == "title":
        return "".join(t.get("plain_text", "") for t in prop.get("title") or [])
    return ""


def read_priority(prop: object) -> str:
    if not isinstance(prop, dict):
        return ""
    kind = prop.get("type")
    if kind == "select":
        return (prop.get("select") or {}).get("name") or ""
    if kind == "status":
        return (prop.get("status") or {}).get("name") or ""
    if kind == "number" and prop.get("number") is not None:
        return str(prop["number"])
    if kind == "rich_text":
        return "".join(t.get("plain_text", "") for t in prop.get("rich_text") or [])
    return ""


def title_from_page(page: dict[str, Any], title_prop_name: str) -> str:
    props = page.get("properties") or {}
    named = props.get(title_prop_name)
    if isinstance(named, dict) and named.get("type") == "title":
        return "".join(t.get("plain_text", "") for t in named.get("title") or []) or "Untitled"
    for prop in props.values():
        if isinstance(prop, dict) and prop.get("type") == "title":
            return "".join(t.get("plain_text", "") for t in prop.get("title") or []) or "Untitled"
    return "Untitled"


def extract_project_brief_fields(page: dict[str, Any]) -> dict[str, Any]:
    props = page.get("properties") or {}
    status = ""
    st = props.get(PROJECT_PROPS["status"])
    if isinstance(st, dict):
        if st.get("type") == "status":
            status = (st.get("status") or {}).get("name") or ""
        elif st.get("type") == "select":
            status = (st.get("select") or {}).get("name") or ""
    deadline = None
    dl = props.get(PROJECT_PROPS["deadline"])
    if isinstance(dl, dict) and dl.get("type") == "date":
        deadline = (dl.get("date") or {}).get("start")
    return {
        "name": title_from_page(page, PROJECT_PROPS["title"]),
        "status": status,
        "priority": read_priority(props.get(PROJECT_PROPS["priority"])),
        "currentFocus": read_rich_text(props.get(PROJECT_PROPS["currentFocus"])),
        "nextAction": read_rich_text(props.get(PROJECT_PROPS["nextAction"])),
        "deadline": deadline,
    }


def notion_status_db_filter(property_name: str, equals: str) -> dict[str, Any]:
    if get_settings().NOTION_STATUS_PROPERTY_KIND == "select":
        return {"property": property_name, "select": {"equals": equals}}
    return {"property": property_name, "status": {"equals": equals}}


def notion_status_update_payload(name: str) -> dict[str, Any]:
    if get_settings().NOTION_STATUS_PROPERTY_KIND == "select":
        return {"select": {"name": name}}
    return {"status": {"name": name}}


def notion_status_clear_payload() -> dict[str, Any]:
    if get_settings().NOTION_STATUS_PROPERTY_KIND == "select":
        return {"select": None}
    return {"status": None}


def is_full_page(obj: object) -> bool:
    return isinstance(obj, dict) and obj.get("object") == "page" and "properties" in obj
